realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -76,64 +81,112 @@ static RealmCacheType valueOf(Class<? extends BaseRealm> clazz) {
      * @return the {@link Realm} or {@link DynamicRealm} instance.
      */
-    static synchronized <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
+    static <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
             Class<E> realmClass) {
-        boolean isCacheInMap = true;
-        RealmCache cache = cachesMap.get(configuration.getPath());
-        if (cache == null) {
-            // Creates a new cache.
-            cache = new RealmCache(configuration);
-            // The new cache should be added to the map later.
-            isCacheInMap = false;
+        RealmCache cache = getCache(configuration.getPath(), true);
+
+        return cache.doCreateRealmOrGetFromCache(configuration, realmClass);
+    }
 
+    private synchronized <E extends BaseRealm> E doCreateRealmOrGetFromCache(RealmConfiguration configuration,
+            Class<E> realmClass) {
+
+        RefAndCount refAndCount = refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+
+        if (getTotalGlobalRefCount() == 0) {
             copyAssetFileIfNeeded(configuration);
-        } else {
-            // Throws the exception if validation failed.
-            cache.validateConfiguration(configuration);
-        }
 
-        RefAndCount refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+            SharedRealm sharedRealm = null;
+            try {
+                sharedRealm = SharedRealm.getInstance(configuration);
+                if (Table.primaryKeyTableNeedsMigration(sharedRealm)) {
+                    sharedRealm.beginTransaction();
+                    if (Table.migratePrimaryKeyTableIfNeeded(sharedRealm)) {
+                        sharedRealm.commitTransaction();
+                    } else {
+                        sharedRealm.cancelTransaction();
+                    }
+                }
 
-        if (refAndCount.globalCount == 0) {
-            SharedRealm sharedRealm = SharedRealm.getInstance(configuration);
-            if (Table.primaryKeyTableNeedsMigration(sharedRealm)) {
-                sharedRealm.beginTransaction();
-                if (Table.migratePrimaryKeyTableIfNeeded(sharedRealm)) {
-                    sharedRealm.commitTransaction();
-                } else {
-                    sharedRealm.cancelTransaction();
+            } finally {
+                if (sharedRealm != null) {
+                    sharedRealm.close();
                 }
             }
-            sharedRealm.close();
+
+            // We are holding the lock, and we can set the invalidated configuration since there is no global ref to it.
+            this.configuration = configuration;
+        } else {
+            // Throws exception if validation failed.
+            validateConfiguration(configuration);
         }
 
         if (refAndCount.localRealm.get() == null) {
```

Comment:
`refAndCount.localRealm.get() != nul`l and `refAndCount.localCount.get() == 0` should never happen, but i see your point, it is more readable if i merge the block 212-220 to line 192 as you mentioned before.